File: topKFrequent.js
function topKFrequent(nums, k){
    let result = []
    let obj = {}

    for(let i=0; i<nums.length; i++){
        let num=nums[i]
        if(obj[num]){
            obj[num]++
        } else {
            obj[num] = 1
        }
    }
    let sorted = Object.entries(obj).sort((a,b)=> b[1]-a[1])
    // The Object.entries() method in JavaScript returns an array of a given object's own[key, value] pairs.
    // .sort() is a JavaScript array method that sorts the elements of an array in-place and returns the array.
    // (a, b) => b[1] - a[1] is an arrow function that is passed to the sort method as a comparator function. This function is used to define the sort order.
    // a and b within the arrow function represent two elements being compared. In your case, since you're sorting an array of arrays, a and b are arrays themselves.
    // b[1] - a[1] means that the sort order is determined by subtracting the second element of array a from the second element of array b.
    for(let i=0; i<k; i++){
        result.push(parseInt(sorted[i][0]))
    }
    return result
}

topKFrequent([1,1,1,2,2,3], 2)

module.exports = topKFrequent;

File: topKFrequent.py
def top_k_frequent(nums, k):
    #empty object to keep track of how many times each integer appears in the array
    #empty array to hold result
    #iterate over array. check for current integer. if it's in the object, increment it's number. if it's not, add it.
    #sort the object key:value pairs (arrays) in descending order
    #iterate from 0 to k and on each iteration, push the current value into the result array

    result = []
    obj = {}

    for num in nums:
        if obj.get(num):
            obj[num]+=1
        else:
            obj[num]=1
    
    sorted_list = sorted(obj.items(), key=lambda x:x[1], reverse=True)
    #obj.items() returns a list of (key, value) pairs from the dictionary obj. 
    #sorted(...) sorts this list. It takes two optional arguments: key, a function that returns the value to sort by, and reverse, a boolean that specifies whether to sort in ascending (the default) or descending order.
    #key=lambda x: x[1] specifies that the sorting should be based on the second element of each pair (the value), and reverse=True specifies that the sorting should be in descending order. 
    for i in range(k):
        result.append(sorted_list[i][0])

    print(result) 

top_k_frequent([1,1,1,2,2,3], 2)